Alignment/include/ActsAlignment/Kernel/Alignment.hpp

```diff
@@ -0,0 +1,513 @@
+  AlignedTransformUpdater alignedTransformUpdater = nullptr;
+
+  // The detector elements to be aligned
+  std::vector<Acts::DetectorElementBase*> alignedDetElements;
+
+  // The alignment tolerance to determine if the alignment is coveraged
+  double averageChi2ONdfCutOff = 0.5;
+
+  // The delta of average chi2/ndf within a couple of iterations to determine if
+  // alignment is converged
+  std::pair<size_t, double> deltaAverageChi2ONdfCutOff = {5, 0.01};
+
+  // The maximum number of iterations to run alignment
+  size_t maxIterations = 5;
+
+  // The alignment mask for different iterations
+  std::map<unsigned int, AlignmentMask> iterationState;
+
+  /// Logger
+  Acts::LoggerWrapper logger;
+};
+
+/// @brief Alignment result struct
+///
+struct AlignmentResult {
+  // The change of alignment parameters
+  Acts::ActsDynamicVector deltaAlignmentParameters;
+
+  // The aligned parameters for detector elements
+  std::unordered_map<Acts::DetectorElementBase*, Acts::Transform3>
+      alignedParameters;
+
+  // The covariance of alignment parameters
+  Acts::ActsDynamicMatrix alignmentCovariance;
+
+  // The avarage chi2/ndf (ndf is the measurement dim)
+  double averageChi2ONdf = std::numeric_limits<double>::max();
+
+  // The delta chi2
+  double deltaChi2 = std::numeric_limits<double>::max();
+
+  // The chi2
+  double chi2 = 0;
+
+  // The measurement dimension from all tracks
+  size_t measurementDim = 0;
+
+  // The alignment degree of freedom
+  size_t alignmentDof = 0;
+
+  // The number of tracks used for alignment
+  size_t numTracks = 0;
+
+  // The indexed alignable surfaces
+  std::unordered_map<const Acts::Surface*, size_t> idxedAlignSurfaces;
+
+  Acts::Result<void> result{Acts::Result<void>::success()};
+};
+
+/// @brief KalmanFitter-based alignment implementation
+///
+/// @tparam fitter_t Type of the fitter class
+template <typename fitter_t>
+struct Alignment {
+  /// Default constructor is deleted
+  Alignment() = delete;
+
+  /// Constructor from arguments
+  Alignment(fitter_t fitter) : m_fitter(std::move(fitter)) {}
+
+  /// @brief evaluate alignment state for a single track
+  ///
+  /// @tparam source_link_t Source link type identifying uncalibrated input
+  /// measurements.
+  /// @tparam start_parameters_t Type of the initial parameters
+  /// @tparam fit_options_t The fit options type
+  ///
+  /// @param gctx The current geometry context object
+  /// @param sourcelinks The fittable uncalibrated measurements
+  /// @param sParameters The initial track parameters
+  /// @param fitOptions The fit Options steering the fit
+  /// @param idxedAlignSurfaces The idxed surfaces to be aligned
+  /// @param alignMask The alignment mask (same for all detector element for the
+  /// moment)
+  ///
+  /// @result The alignment state for a single track
+  template <typename source_link_t, typename start_parameters_t,
+            typename fit_options_t>
+  Acts::Result<detail::TrackAlignmentState> evaluateTrackAlignmentState(
```

Comment:
Is this result used somewhere outside the alginment code? (just asking for the `detail` namespace)

Reply:
No, it's used only for the alignment code.

Reply:
Ok.